src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -2633,6 +2651,41 @@ namespace QgsWms
     return layers;
   }
 
+  QgsMapLayer *QgsRenderer::createExternalWMSLayer( const QString &externalLayerId ) const
+  {
+    QgsMapLayer *wmsLayer = 0;
+    QgsDataSourceUri wmsUri;
+
+    QgsServerRequest::Parameters::const_iterator it = mParameters.lowerBound( externalLayerId.toUpper() + ":" );
+    while ( it != mParameters.constEnd() )
+    {
+      if ( !it.key().startsWith( externalLayerId.toUpper() + ":" ) )
+      {
+        break;
+      }
+
+      QString paramKey = it.key();
+      paramKey.remove( 0, externalLayerId.size() + 1 );
+      wmsUri.setParam( paramKey.toLower(), it.value() );
+      ++it;
+    }
```

Comment:
The wmsUri construction has to be part of the QgsWmsParameters class.

Reply:
@mhugent any answer to @rldhont 's comment ?

Reply:
@mhugo it has been done

Reply:
I agree with rldhont's comment and it is adressed in commit https://github.com/qgis/QGIS/pull/5396/commits/c99388804a2a32608fb239fbb29b25a6952f5177

Reply:
oops, sorry, my bad !